core/src/main/java/io/grpc/CallOptions.java

```diff
@@ -295,22 +295,31 @@ public String toString() {
   }
 
   /**
-   * Set a custom option.  Any existing value for the key overridden.
-   * 
+   * Sets a custom option. Any existing value for the key is overwritten.
+   *
    * @param key The option key
    * @param value The option value.
    */
   @ExperimentalApi("https://github.com/grpc/grpc-java/issues/1869")
   public <T> CallOptions withOption(Key<T> key, T value) {
     Preconditions.checkNotNull(key);
     Preconditions.checkNotNull(value);
-      
+
     CallOptions newOptions = new CallOptions(this);
-    newOptions.customOptions = new Object[customOptions.length + 1][2];
-    newOptions.customOptions[0] = new Object[] { key, value};
-    if (customOptions.length > 0) {
-      System.arraycopy(customOptions, 0, newOptions.customOptions, 1, customOptions.length);
+    // Overwrite the value if the key exists.
```

Comment:
Since you have bench marks, you might consider also try:

```
idx = -1
for (...) {
  if (key == arr[i]) {
    idx = i;
    break;
  }
}

newArr = new Object[arr.length + idx == -1 ? 0 : 1][2]
System.arrayCopy
if (idx == -1) { 
  append
} else {
  replace
}
```

That way you only have one possible call to array copy